apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/extract/s3/DataSetQueue.java

```diff
@@ -1,266 +1,358 @@
-import software.amazon.awssdk.core.exception.SdkServiceException;
+import javax.annotation.Nonnull;
+import lombok.AllArgsConstructor;
+
+/**
+ * Represents a queue of data sets (manifest file plus entry files) by combining the current
+ * contents of an S3 bucket and records in the database reflecting the status of those files.
+ */
+@AllArgsConstructor
+public class DataSetQueue implements AutoCloseable {
+  /**
+   * Name of S3 meta data field used by CCW to communicate an expected MD5 checksum value for every
+   * file they upload to the S3 bucket for processing.
+   */
+  public static final String MD5_CHECKSUM_META_DATA_FIELD = "md5chksum";
+
+  /** Name used for read list of eligible manifests timer. */
+  static final String TIMER_READ_MANIFESTS =
+      MetricRegistry.name(DataSetQueue.class, "readEligibleManifests");
+
+  /** Name used for download one manifest timer. */
+  static final String TIMER_DOWNLOAD_MANIFEST =
+      MetricRegistry.name(DataSetQueue.class, "downloadManifest");
+
+  /** Name used for download one manifest entry timer. */
+  static final String TIMER_DOWNLOAD_ENTRY =
+      MetricRegistry.name(DataSetQueue.class, "downloadEntry");
+
+  /** Name used for updating one manifest in database timer. */
+  static final String TIMER_MANIFEST_DB_UPDATE =
+      MetricRegistry.name(DataSetQueue.class, "updateManifestInDb");
 
-/** Represents and manages the queue of data sets in S3 to be processed. */
-public final class DataSetQueue {
-  private static final Logger LOGGER = LoggerFactory.getLogger(DataSetQueue.class);
+  /** Statuses that indicate that processing is allowed on a {@link S3ManifestFile}. */
+  private static final Set<S3ManifestFile.ManifestStatus> STARTABLE_MANIFEST_STATUSES =
+      Set.of(S3ManifestFile.ManifestStatus.DISCOVERED, S3ManifestFile.ManifestStatus.STARTED);
+
+  /** Statuses that indicate that processing is allowed on a {@link S3DataFile}. */
+  private static final Set<S3DataFile.FileStatus> STARTABLE_ENTRY_STATUSES =
+      Set.of(S3DataFile.FileStatus.DISCOVERED, S3DataFile.FileStatus.STARTED);
+
+  /** Used to access current time. */
+  private final Clock clock;
 
   /** The metric registry. */
   private final MetricRegistry appMetrics;
 
-  /** The extraction options. */
-  private final ExtractionOptions options;
+  /** Used to track S3 files in the database. */
+  private final S3ManifestDbDao s3Records;
 
-  /** The S3 task manager. */
-  private final S3TaskManager s3TaskManager;
+  /** Used to download files from S3 to a local cache for processing. */
+  private final S3FileManager s3Files;
 
   /**
-   * The {@link DataSetManifest}s waiting to be processed, ordered by their {@link
-   * DataSetManifestId} in ascending order such that the first element represents the {@link
-   * DataSetManifest} that should be processed next.
+   * Used only for the soon to be obsolete S3 file move task.
+   *
+   * <p>TODO Remove this when file moves are no longer necessary.
```

Comment:
Another place to maybe toss a story number so we can find these all when its time.